From 3cf0eee6a88f77c18813a662a8740840704023fd Mon Sep 17 00:00:00 2001
From: Jacky Chou <jacky_chou@aspeedtech.com>
Date: Wed, 6 May 2026 11:06:17 +0800
Subject: [PATCH 09/13] net: dwc_eth_xgmac: add 10G support and update ASPEED
 config

Add xgmac_set_xgmii_speed_10g() and handle SPEED_10000 in
xgmac_adjust_link() to configure the MAC for 10G XGMII mode.

Update the ASPEED-specific config: disable RXQ0 DCB mode
(RXQ0EN_NOT_ENABLED), switch MDIO clock ratio to CR_300_350, and
widen the AXI bus to 128-bit. Select PHY_DW_XPCS from
DWC_ETH_XGMAC_ASPEED so the XPCS PHY is always built in.

Signed-off-by: Jacky Chou <jacky_chou@aspeedtech.com>
Change-Id: I47ecce51b55b254782e365d51186db4f8c2dfba8
Upstream-Status: Backport [SDK v11.03 migration]
---
 drivers/net/Kconfig                |  1 +
 drivers/net/dwc_eth_xgmac.c        | 18 ++++++++++++++++++
 drivers/net/dwc_eth_xgmac_aspeed.c |  6 +++---
 3 files changed, 22 insertions(+), 3 deletions(-)

diff --git a/drivers/net/Kconfig b/drivers/net/Kconfig
index 34b9b4b045..d6e2becf89 100644
--- a/drivers/net/Kconfig
+++ b/drivers/net/Kconfig
@@ -204,6 +204,7 @@ config DWC_ETH_XGMAC
 config DWC_ETH_XGMAC_ASPEED
 	bool "Synopsys DWC Ethernet XGMAC device support for ASPEED"
 	select DWC_ETH_XGMAC
+	select PHY_DW_XPCS
 	help
 	  The Synopsys Designware Ethernet XGMAC IP block with specific
 	  configuration used in ASPEED chip.
diff --git a/drivers/net/dwc_eth_xgmac.c b/drivers/net/dwc_eth_xgmac.c
index aaac30296a..9e762987d8 100644
--- a/drivers/net/dwc_eth_xgmac.c
+++ b/drivers/net/dwc_eth_xgmac.c
@@ -328,6 +328,20 @@ static int xgmac_set_gmii_speed(struct udevice *dev)
 	return 0;
 }
 
+static int xgmac_set_xgmii_speed_10g(struct udevice *dev)
+{
+	struct xgmac_priv *xgmac = dev_get_priv(dev);
+	u32 val;
+
+	debug("%s(dev=%p):\n", __func__, dev);
+
+	/* FPGA -> 5G */
+	val = XGMAC_MAC_CONF_SS_10G_XGMII << XGMAC_MAC_CONF_SS_SHIFT;
+	writel(val, &xgmac->mac_regs->tx_configuration);
+
+	return 0;
+}
+
 static int xgmac_set_mii_speed_100(struct udevice *dev)
 {
 	struct xgmac_priv *xgmac = dev_get_priv(dev);
@@ -372,6 +386,10 @@ static int xgmac_adjust_link(struct udevice *dev)
 	}
 
 	switch (xgmac->phy->speed) {
+	case SPEED_10000:
+		en_calibration = false;
+		ret = xgmac_set_xgmii_speed_10g(dev);
+		break;
 	case SPEED_1000:
 		en_calibration = true;
 		ret = xgmac_set_gmii_speed(dev);
diff --git a/drivers/net/dwc_eth_xgmac_aspeed.c b/drivers/net/dwc_eth_xgmac_aspeed.c
index a7a64455d0..8ca1571766 100644
--- a/drivers/net/dwc_eth_xgmac_aspeed.c
+++ b/drivers/net/dwc_eth_xgmac_aspeed.c
@@ -134,9 +134,9 @@ static struct xgmac_ops xgmac_aspeed_ops = {
 struct xgmac_config __maybe_unused xgmac_aspeed_config = {
 	.reg_access_always_ok = false,
 	.swr_wait = 50,
-	.config_mac = XGMAC_MAC_RXQ_CTRL0_RXQ0EN_ENABLED_DCB,
-	.config_mac_mdio = XGMAC_MAC_MDIO_ADDRESS_CR_350_400,
-	.axi_bus_width = XGMAC_AXI_WIDTH_64,
+	.config_mac = XGMAC_MAC_RXQ_CTRL0_RXQ0EN_NOT_ENABLED,
+	.config_mac_mdio = XGMAC_MAC_MDIO_ADDRESS_CR_300_350,
+	.axi_bus_width = XGMAC_AXI_WIDTH_128,
 	.interface = dev_read_phy_mode,
 	.ops = &xgmac_aspeed_ops
 };
-- 
2.34.1

